### ISSUE TITLE
Streamplot ignores rightmost column and topmost row of velocity data

### ISSUE THREAD

**Issue by pmackenz** [matplotlib/matplotlib#11452]
### Bug report

**Bug summary**

streamplot ignores the rightmost column (and topmost row) in U and V and replaces them by a copy of the second to last column (row).  This yields wrong stream lines in the rightmost column and topmost row of cells.  The issue is most noticeable on coarse grids. (see example plots)

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt

x = np.linspace(-1, 1, 3)
y = np.linspace(0, 1, 2)

U = np.array([[.2,.2,.2],[.2,1,.2]])
V = np.array([[-1,2,-1],[-1,0.,-1]])

plt.streamplot(x,y, U,V)
plt.quiver(x,y, U,V)
plt.xlim(-1.5,1.5)
plt.ylim(-0.5,1.5)
plt.savefig('StreamPlotIssue1.png')
plt.close()

x = np.linspace(-1, 1, 5)
y = np.linspace(0, 1, 4)

U = np.array([[.2,.2,.2,.2,.2],[.2,.2,.2,.2,.2],[.2,.2,.2,.2,.2],[.2,1.,1.,1,.2]])
V = np.array([[-1,.5,2,.5,-1],[-1,.5,2,.5,-1],[-1,.5,2,.5,-1],[-1,0.,0.,0.,-1]])

plt.streamplot(x,y, U,V)
plt.quiver(x,y, U,V)
plt.xlim(-1.5,1.5)
plt.ylim(-0.5,1.5)
plt.savefig('StreamPlotIssue2.png')
plt.close()
```

**Actual outcome**

The first image uses just 2 cells (2 in x, 1 in y direction).  The field should create an antisymmetric wave.  The left cell is OK, the right one is garbage. (The example is chosen such that the top row issue doesn't show too badly).  The overlaid quiver plot shows the correct velocity.

![streamplotissue1](https://user-images.githubusercontent.com/15879316/41511048-6b02fbde-7224-11e8-83f6-509293c22984.png)

The second image uses 4 cells in x and 3 cells in y-direction for the same velocity field.  Quiver shows the correct velocity as overlay.  The rightmost column and top row of cells show the wrong streamlines.

![streamplotissue2](https://user-images.githubusercontent.com/15879316/41511052-748d1e50-7224-11e8-8615-9c851037d028.png)

For quick reference: the variables used
```
# image 1
#x:
[-1.  0.  1.]
# y:
[0. 1.]
# U:
[[0.2 0.2 0.2]
 [0.2 1.  0.2]]
# V:
[[-1.  2. -1.]
 [-1.  0. -1.]]

# image 2

#x:
[-1.  -0.5  0.   0.5  1. ]
# y:
[0.         0.33333333 0.66666667 1.        ]
# U:
[[0.2 0.2 0.2 0.2 0.2]
 [0.2 0.2 0.2 0.2 0.2]
 [0.2 0.2 0.2 0.2 0.2]
 [0.2 1.  1.  1.  0.2]]
# V:
[[-1.   0.5  2.   0.5 -1. ]
 [-1.   0.5  2.   0.5 -1. ]
 [-1.   0.5  2.   0.5 -1. ]
 [-1.   0.   0.   0.  -1. ]]
```

**Expected outcome**

A streamline plot consistent with velocity shown in quiver overlay.

**Matplotlib version**

  * Operating system:
MacOS High Sierra (10.13.5)
  * Matplotlib version: 
matplotlib-2.1.2-py36h6d6146d_0
  * Matplotlib backend (`print(matplotlib.get_backend())`):
MacOSX
  * Python version:
3.6
  * Jupyter version (if applicable):
n/a
  * Other libraries: 
numpy

Python and matplotlib installed using anaconda3 package

**Comment by pharshalp** [comment]
The interpolation (in the case of streamplot) might be failing because of fewer number of grid values specified (the V component changes sign and only 3 points are used to capture this change).

If I increase the number of points to 9 (adding evenly spaced values) then the results seems to be correct... 

![streamplotissue_more_points](https://user-images.githubusercontent.com/15175620/41514505-7d52a766-7275-11e8-9ede-8127f6dd018e.png)

`import numpy as np
import matplotlib.pyplot as plt

import numpy as np
import matplotlib.pyplot as plt

x = np.linspace(-1, 1, 9)
y = np.linspace(0, 1, 2)

U = np.array([[.2,.2,.2,.2,.2,.2,.2,.2,.2],
	                [.2,0.4,0.6,0.8,1,0.8,0.6,0.4,.2]])
V = np.array([[-1,-0.25, 0.5, 1.25, 2, 1.25, 0.5, -0.25, -1],
	               [-1,-0.75,-0.5,-0.25,0.,0.25, 0.5, 0.75, -1]])

plt.streamplot(x,y, U,V)
plt.quiver(x,y, U,V)
plt.xlim(-1.5,1.5)
plt.ylim(-0.5,1.5)
plt.savefig('StreamPlotIssue_more_points.png')
plt.close()`

**Comment by jklymak** [comment]
A streamline is defined as the integral of the crossstream velocity, so I wouldn’t be surprised that there are edge effects for a digital representation.  If you need a smoother stream plot interpolate the data first.

**Comment by jklymak** [comment]
Any finite representation will have gridscale inaccuracies compared to a formal definition.  If you have a proposed algorithm that deals with those inaccuracies in a different way than the matplotlib implementation of streamline a pull request would receive due attention.

**Comment by pmackenz** [comment]
Pull Request with bug fix submitted :)

Here the new, now correct, output of the above sample code:
![streamplotissue1](https://user-images.githubusercontent.com/15879316/41553604-1961c7f6-72e7-11e8-8815-90fc4540eb5b.png)
![streamplotissue2](https://user-images.githubusercontent.com/15879316/41553612-1dbab218-72e7-11e8-80b4-1f9a29a55e01.png)

**Comment by pmackenz** [comment]
... and here the correct streamlines for your data set:
![streamplotissue_more_points](https://user-images.githubusercontent.com/15879316/41553727-77759bb0-72e7-11e8-89fa-52ba8278874b.png)

**Comment by ImportanceOfBeingErnest** [comment]
Link to potential fix: #11455

**Comment by efiring** [comment]
closed by #11455
